language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -40,25 +37,16 @@ newtype DynamicBuiltinNameTypes = DynamicBuiltinNameTypes
 -- | Configuration of the type checker.
 data TypeConfig = TypeConfig
     { _typeConfigNormalize           :: Bool
-      -- ^ Whether to normalize type annotations
+      -- ^ Whether to normalize type annotations.
     , _typeConfigDynBuiltinNameTypes :: DynamicBuiltinNameTypes
-    }
-
-newtype TypeCheckSt = TypeCheckSt
-    { _gas :: Natural
-    }
-
-data TypeCheckCfg = TypeCheckCfg
-    { _cfgGas        :: Natural     -- ^ Gas to be provided to the typechecker
-    , _cfgTypeConfig :: TypeConfig
+    , _typeConfigGas                 :: Maybe Natural -- ^ The upper limit on the length of type
```

Comment:
So here it's clear what `Maybe Natural` means, but not in the other multiple places. Why not define a data type for this?

```haskell
data TransNatural
    = Natural Natural
    | Omega
```

or something like that. This makes the more structured and one can just jump to the definition of `TransNatural` and read the comment. While having an uncommented `Maybe Natural` at hands (like in `parseTypecheck` or `NormalizeTypeM`, for example) is not helpful.

Reply:
I'm not a fan, mainly because this would go out of our way to not use stock types for no discernible reason.

Reply:
I'd be happy with `newtype TransNatural = TransNatural (Maybe Natural)`. There is a very good reason to do this: person seeing `Maybe Natural` as the type of an argument has no idea what this all is about. I think we even should have `newtype GasInit = GasInit (Maybe Gas)`. `Maybe Natural` is way too underspecified.